File: backend/shared/src/shared/schemas/core/team_invitation.py
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Optional

import humps
from pydantic import TypeAdapter

from shared.schemas.dto.team_invitation import (
    EnrichedTeamInvitationDTO,
    TeamInvitationDTO,
)


class TeamInvitationStatus(Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    REJECTED = "rejected"
    EXPIRED = "expired"


class TeamInvitationType(Enum):
    MEMBER = "member"


# pylint: disable=too-many-instance-attributes
@dataclass
class TeamInvitation:
    id: str
    team_id: str
    email: str
    type: TeamInvitationType
    worker_id: Optional[str]
    token: str
    status: TeamInvitationStatus
    created_at: datetime
    expires_at: datetime
    last_sent_at: Optional[datetime] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    created_by: Optional[str] = None

    def to_dto(self) -> TeamInvitationDTO:
        data = asdict(self)
        data["type"] = self.type.value
        data["status"] = self.status.value
        data["created_at"] = self.created_at.timestamp()
        data["expires_at"] = self.expires_at.timestamp()
        data["last_sent_at"] = (
            self.last_sent_at.timestamp() if self.last_sent_at else None
        )
        as_dict = humps.camelize(data)
        validator = TypeAdapter(TeamInvitationDTO)
        return validator.validate_python(as_dict)

    @classmethod
    def from_dto(cls, data: TeamInvitationDTO) -> "TeamInvitation":
        data_dict = humps.decamelize(data.model_dump())
        data_dict["type"] = TeamInvitationType(data_dict["type"])
        data_dict["status"] = TeamInvitationStatus(data_dict["status"])
        data_dict["created_at"] = datetime.fromtimestamp(
            data_dict["created_at"], tz=timezone.utc
        )
        data_dict["expires_at"] = datetime.fromtimestamp(
            data_dict["expires_at"], tz=timezone.utc
        )
        data_dict["last_sent_at"] = (
            datetime.fromtimestamp(data_dict["last_sent_at"], tz=timezone.utc)
            if data_dict["last_sent_at"]
            else None
        )
        return cls(**data_dict)


@dataclass
class EnrichedTeamInvitation:
    id: str
    team_id: str
    team_name: str  # enriched
    email: str
    type: TeamInvitationType
    worker_id: Optional[str]
    token: str
    status: TeamInvitationStatus
    created_at: datetime
    expires_at: datetime
    last_sent_at: Optional[datetime] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    created_by: Optional[str] = None
    creator_name: Optional[str] = None  # enriched

    def to_dto(self) -> EnrichedTeamInvitationDTO:
        data = asdict(self)
        data["type"] = self.type.value
        data["status"] = self.status.value
        data["created_at"] = self.created_at.timestamp()
        data["expires_at"] = self.expires_at.timestamp()
        data["last_sent_at"] = (
            self.last_sent_at.timestamp() if self.last_sent_at else None
        )
        as_dict = humps.camelize(data)
        validator = TypeAdapter(EnrichedTeamInvitationDTO)
        return validator.validate_python(as_dict)

    @classmethod
    def from_dto(cls, data: EnrichedTeamInvitationDTO) -> "EnrichedTeamInvitation":
        data_dict = humps.decamelize(data.model_dump())
        data_dict["type"] = TeamInvitationType(data_dict["type"])
        data_dict["status"] = TeamInvitationStatus(data_dict["status"])
        data_dict["created_at"] = datetime.fromtimestamp(
            data_dict["created_at"], tz=timezone.utc
        )
        data_dict["expires_at"] = datetime.fromtimestamp(
            data_dict["expires_at"], tz=timezone.utc
        )
        data_dict["last_sent_at"] = (
            datetime.fromtimestamp(data_dict["last_sent_at"], tz=timezone.utc)
            if data_dict["last_sent_at"]
            else None
        )
        return cls(**data_dict)

File: frontend/src/types/team-invitation.ts
import dayjs from 'dayjs';

export enum TeamInvitationStatus {
  PENDING = 'pending',
  ACCEPTED = 'accepted',
  REJECTED = 'rejected',
  EXPIRED = 'expired',
}
export enum TeamInvitationType {
  MEMBER = 'member',
}
export type TeamInvitationT = {
  id: string;
  teamId: string;
  firstName: string | null;
  lastName: string | null;
  email: string;
  type: TeamInvitationType;
  workerId: string | null;
  token: string;
  status: TeamInvitationStatus;
  createdBy: string | null;
  createdAt: dayjs.Dayjs;
  expiresAt: dayjs.Dayjs;
  lastSentAt: dayjs.Dayjs | null;
};

export type EnrichedTeamInvitationT = {
  id: string;
  teamId: string;
  firstName: string | null;
  lastName: string | null;
  email: string;
  type: TeamInvitationType;
  workerId: string | null;
  token: string;
  status: TeamInvitationStatus;
  createdBy: string | null;
  createdAt: dayjs.Dayjs;
  expiresAt: dayjs.Dayjs;
  lastSentAt: dayjs.Dayjs | null;
  teamName: string;
  creatorName: string | null;
};

export const toTeamInvitationT = (data: any): TeamInvitationT => {
  return {
    ...data,
    createdAt: dayjs.unix(data.createdAt).utc(),
    expiresAt: dayjs.unix(data.expiresAt).utc(),
    lastSentAt: data.lastSentAt ? dayjs.unix(data.lastSentAt).utc() : null,
  };
};

export const fromTeamInvitationT = (data: TeamInvitationT): any => {
  return {
    ...data,
    createdAt: data.createdAt.unix(),
    expiresAt: data.expiresAt.unix(),
    lastSentAt: data.lastSentAt ? data.lastSentAt.unix() : null,
  };
};

export const toEnrichedTeamInvitationT = (data: any): EnrichedTeamInvitationT => {
  return {
    ...data,
    createdAt: dayjs.unix(data.createdAt).utc(),
    expiresAt: dayjs.unix(data.expiresAt).utc(),
    lastSentAt: data.lastSentAt ? dayjs.unix(data.lastSentAt).utc() : null,
  };
};

export const fromEnrichedTeamInvitationT = (data: EnrichedTeamInvitationT): any => {
  return {
    ...data,
    createdAt: data.createdAt.unix(),
    expiresAt: data.expiresAt.unix(),
    lastSentAt: data.lastSentAt ? data.lastSentAt.unix() : null,
  };
};
